truffleruby/src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -415,15 +415,9 @@ public DynamicObject wakeup(DynamicObject thread,
 
         @Specialization(guards = {"isRubyProc(unblocker)", "isRubyProc(runner)"})
         public Object unblock(VirtualFrame frame, DynamicObject thread, DynamicObject unblocker, DynamicObject runner) {
-            Layouts.THREAD.setUnblocker(thread, unblocker);
-            Layouts.THREAD.setStatus(thread, ThreadStatus.SLEEP);
-
-            try {
-                return yield(frame, runner);
-            } finally {
-                Layouts.THREAD.setUnblocker(thread, null);
-                Layouts.THREAD.setStatus(thread, ThreadStatus.RUN);
-            }
+            return getContext().getThreadManager().runUntilResult(this,
```

Comment:
This really depends on the semantics of `rb_thread_call_without_gvl()`.
The blockingAction here will not be retried due to never returning null as it cannot (yield doesn't return null), and it does not throw an InterruptedExeption (the InterruptedExeption is handled by the caller of the method which can throw it, not here).

So I think `runUntilResult` is correct here.